Week-4/assignment-1.js

```diff
@@ -0,0 +1,122 @@
+const STATE = {
+  FULFILLED: "fulfilled",
+  REJECTED: "rejected",
+  PENDING: "pending"
+};
+class Promise {
+  constructor(cb) {
+    this.value;
+    this.state = STATE.PENDING;
+    this.thenCbs = [];
+    this.catchCbs = [];
+
+    this.runCallBacks = () => {
+      if (this.state === STATE.FULFILLED) {
+        this.thenCbs.forEach((callBack) => {
+          callBack(this.value);
+        });
+        /* We need to clear the then callbacks array since if .then() is called 
+          multiple times the old stored callbacks would also get executed 
+          along with new one's  */
+        this.thenCbs = [];
+      }
+
+      if (this.state === STATE.REJECTED) {
+        this.catchCbs.forEach((callBack) => {
+          callBack(this.value);
+        });
+        this.catchCbs = [];
+      }
+    };
+
+    const onSuccess = (value) => {
+      if (this.state !== STATE.PENDING) {
+        return;
+      } else {
+        this.value = value;
+        this.state = STATE.FULFILLED;
+        this.runCallBacks();
+      }
+    };
+    const onFail = (value) => {
+      if (this.state !== STATE.PENDING) {
+        return;
+      } else {
+        this.value = value;
+        this.state = STATE.REJECTED;
+        this.runCallBacks();
+      }
+    };
+
+    try {
+      cb(onSuccess, onFail);
+    } catch (err) {
+      onFail(err);
+    }
+  }
+
+  then(thenCb, catchCb) {
+    return new Promise((resolve, reject) => {
+      this.thenCbs.push((result) => {
+        if (thenCb == null) {
+          resolve(result);
+          return;
+        }
+        try {
+          resolve(thenCb(result));
+        } catch (error) {
+          reject(error);
+        }
+      });
+      this.catchCbs.push((result) => {
+        if (catchCb == null) {
+          reject(result);
+          return;
+        }
+
+        try {
+          resolve(catchCb(result));
+        } catch (error) {
+          reject(error);
+        }
+      });
+      this.runCallBacks();
+    });
+  }
+
+  catch(cb) {
+    return this.then(undefined, cb);
+  }
+
+  finally(cb) {
+    return this.then((result) => {
+      cb();
+      return result;
+    });
+  }
+}
+
+const getNumber = () => {
+  let randomNum = Math.floor(Math.random() * 1000);
+
+  return new Promise((resolve, reject) => {
```

Comment:
Also verify if its picking from javascript promise or ur custom promise.. since same name is used...